src/org/skypro/skyshop/searchEngine/SearchEngine.java

```diff
@@ -3,38 +3,48 @@
 import org.skypro.skyshop.BestResultNotFound;
 import org.skypro.skyshop.Searchable;
 
- hw-6
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Map;
-import java.util.TreeMap;
+import java.util.*;
 
 public class SearchEngine {
-    private final List<Searchable> searchables = new ArrayList<>();
+    private final Set <Searchable> searchables = new HashSet<>();
+
+
+    private static class SearchableComparator implements Comparator<Searchable> {
+        @Override
+        public int compare(Searchable s1, Searchable s2) {
+            int lengthCompare = Integer.compare(s2.getSearchableName().length(), s1.getSearchableName().length());
+            if (lengthCompare == 0) {
+                return s1.getSearchableName().compareTo(s2.getSearchableName());
+            }
+            return lengthCompare;
+        }
+    }
+
+    private static final Comparator<Searchable> SEARCHABLE_COMPARATOR = new SearchableComparator();
 
     public void add(Searchable searchable) {
         searchables.add(searchable);
     }
 
-    public Map<String, Searchable> search(String query) {
-        Map<String, Searchable> results = new TreeMap<>();
+    public TreeSet< Searchable> search(String query) {
```

Comment:
```suggestion
    public Set< Searchable> search(String query) {
```
помним про полиморфизм